userspace/libpman/src/ringbuffer.c

```diff
@@ -261,8 +262,14 @@ static inline void *ringbuf__get_first_ring_event(struct ring *r, int pos)
 	 */
 	if(g_state.cons_pos[pos] >= g_state.prod_pos[pos])
 	{
+		// We try to increment the producer and continue. It is likely that the producer
+		// has produced new events on this CPU and these events could have a timestamp
+		// lowest than all the other events in the other buffers.
 		g_state.prod_pos[pos] = smp_load_acquire(r->producer_pos);
-		return NULL;
+		if(g_state.cons_pos[pos] >= g_state.prod_pos[pos])
```

Comment:
Does this handle wraparounds? In a linear space (offsets going from 0 to infinity), `>` would be impossible except for a grave bug, right?

But if the positions wrap around like in kmod, it's totally possible for the consumer pointer to be ahead of the producer one:

```
|---------------------------|
 #########        ##########
          ^ producer
                  ^ consumer
```

(the `#` characters represent not yet consumed events)

Reply:
> Does this handle wraparounds? In a linear space (offsets going from 0 to infinity), > would be impossible except for a grave bug, right?

Nope, they don't wrap around, so yes the `>` is useless. let's remove it! Thanks

Reply:
Just double-checked with libbpf code, `prod` and `cons` are 2 unsigned long that grow to infinity

```cpp
size_t ring__avail_data_size(const struct ring *r)
{
	unsigned long cons_pos, prod_pos;

	cons_pos = ring__consumer_pos(r);
	prod_pos = ring__producer_pos(r);
	return prod_pos - cons_pos;
}
```

Reply:
done :)